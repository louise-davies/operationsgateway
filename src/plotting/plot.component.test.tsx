import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { FullScalarChannelMetadata, Record, ScalarChannel } from '../app.types';
import ConnectedPlot, {
  ConnectedPlotProps,
  getFormattedAxisData,
  formatTooltipLabel,
} from './plot.component';
import {
  testRecords,
  renderComponentWithProviders,
  generateRecord,
  testChannels,
} from '../setupTests';

const testScalarChannels = testChannels as FullScalarChannelMetadata[];

jest.mock('victory', () => {
  const originalModule = jest.requireActual('victory');

  return {
    __esModule: true,
    ...originalModule,
    VictoryChart: (props) => (
      // @ts-ignore
      <mock-VictoryChart
        {...props}
        scale={JSON.stringify(props.scale, null, 2)}
      />
    ),
    VictoryZoomContainer: (props) => (
      // @ts-ignore
      <mock-VictoryZoomContainer {...props} />
    ),
    VictoryTheme: (props) => (
      // @ts-ignore
      <mock-VictoryTheme {...props} />
    ),
    VictoryScatter: (props) => (
      // @ts-ignore
      <mock-VictoryScatter
        {...props}
        data={JSON.stringify(props.data, null, 2)}
      />
    ),
    VictoryLine: (props) => (
      // @ts-ignore
      <mock-VictoryLine {...props} data={JSON.stringify(props.data, null, 2)} />
    ),
    VictoryLabel: (props) => (
      // @ts-ignore
      <mock-VictoryLabel {...props} />
    ),
    VictoryLegend: (props) => (
      // @ts-ignore
      <mock-VictoryLegend
        {...props}
        data={JSON.stringify(props.data, null, 2)}
      />
    ),
  };
});

describe('Plot component', () => {
  const mockVictoryChart = jest.fn();
  const mockVictoryZoomContainer = jest.fn();
  const mockVictoryTheme = jest.fn();
  const mockVictoryScatter = jest.fn();
  const mockVictoryLine = jest.fn();
  const mockVictoryLabel = jest.fn();
  const mockVictoryLegend = jest.fn();

  const testData: unknown[] = [
    {
      label: 'Test',
      data: [
        { x: 1, y: 1 },
        { x: 2, y: 2 },
        { x: 3, y: 3 },
      ],
    },
  ];

  beforeEach(() => {
    jest.resetModules();
    jest.doMock('victory', () => ({
      VictoryChart: jest.fn((props) => {
        mockVictoryChart(props);
        // @ts-ignore
        return <mock-VictoryChart {...props} />;
      }),
      VictoryZoomContainer: jest.fn((props) => {
        mockVictoryZoomContainer(props);
        // @ts-ignore
        return <mock-VictoryZoomContainer {...props} />;
      }),
      VictoryTheme: jest.fn((props) => {
        mockVictoryTheme(props);
        // @ts-ignore
        return <mock-VictoryTheme {...props} />;
      }),
      VictoryScatter: jest.fn((props) => {
        mockVictoryScatter(props);
        // @ts-ignore
        return <mock-VictoryScatter {...props} />;
      }),
      VictoryLine: jest.fn((props) => {
        mockVictoryLine(props);
        // @ts-ignore
        return <mock-VictoryLine {...props} />;
      }),
      VictoryLabel: jest.fn((props) => {
        mockVictoryLabel(props);
        // @ts-ignore
        return <mock-VictoryLabel {...props} />;
      }),
      VictoryLegend: jest.fn((props) => {
        mockVictoryLegend(props);
        // @ts-ignore
        return <mock-VictoryLegend {...props} />;
      }),
      VictoryTooltip: jest.fn((props) => {
        // @ts-ignore
        return <mock-VictoryTooltip {...props} />;
      }),
    }));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders a scatter plot with the correct elements passed the correct props', () => {
    const { Plot } = require('./plot.component');

    render(
      <Plot
        data={testData}
        title="scatter plot"
        type="scatter"
        XAxisSettings={{ scale: 'time' }}
        YAxesSettings={{ scale: 'linear' }}
      />
    );

    expect(mockVictoryChart).toHaveBeenCalledWith(
      expect.objectContaining({
        scale: { x: 'time', y: 'linear' },
      })
    );
    expect(mockVictoryLabel).toHaveBeenCalledWith(
      expect.objectContaining({
        text: 'scatter plot',
      })
    );
    expect(mockVictoryLegend).toHaveBeenCalled();
    expect(mockVictoryScatter).toHaveBeenCalledWith(
      expect.objectContaining({
        data: testData,
      })
    );
    expect(mockVictoryLine).not.toHaveBeenCalled();
  });

  it('renders a line plot with the correct elements passed the correct props', () => {
    const { Plot } = require('./plot.component');

    render(
      <Plot
        data={testData}
        title="line plot"
        type="line"
        XAxisSettings={{ scale: 'linear' }}
        YAxesSettings={{ scale: 'log' }}
      />
    );

    expect(mockVictoryChart).toHaveBeenCalledWith(
      expect.objectContaining({
        scale: { x: 'linear', y: 'log' },
      })
    );
    expect(mockVictoryLabel).toHaveBeenCalledWith(
      expect.objectContaining({
        text: 'line plot',
      })
    );
    expect(mockVictoryLegend).toHaveBeenCalled();
    expect(mockVictoryScatter).toHaveBeenCalledWith(
      expect.objectContaining({
        data: testData,
      })
    );
    expect(mockVictoryLine).toHaveBeenCalledWith(
      expect.objectContaining({
        data: testData,
      })
    );
  });

  it('redraws the plot in response to resize events', () => {
    const { Plot } = require('./plot.component');

    render(
      <Plot
        data={testData}
        title="Test"
        type="scatter"
        XAxisSettings={{ scale: 'time' }}
        YAxesSettings={{ scale: 'linear' }}
      />
    );

    expect(mockVictoryChart).toHaveBeenCalledTimes(1);

    fireEvent(window, new Event('resize OperationsGateway Plot - Test'));

    // aka it rerenders (it does it twice as redraw is set to true and then reset to false again)
    expect(mockVictoryChart).toHaveBeenCalledTimes(3);
  });
});

describe('formatTooltipLabel function', () => {
  it('formats timestamp correctly', () => {
    const label = 1640995200000;
    const result = formatTooltipLabel(label, 'time');
    expect(result).toEqual('2022-01-01 00:00:00:000');
  });

  it('returns the original label if scale is not time', () => {
    const label = 123456;
    const result = formatTooltipLabel(label, 'linear');
    expect(result).toEqual(label);
  });
});

describe('getFormattedAxisData function', () => {
  let testRecord: Record;

  beforeEach(() => {
    // record with num = 3 creates a record with a scalar channel called test_3
    // this corresponds with scalar metadata channel test_3 in testChannels variable
    testRecord = generateRecord(3);
  });

  it('formats timestamp correctly', () => {
    const unixTimestamp = Math.floor(
      new Date(testRecord.metadata.timestamp).getTime()
    );

    const result = getFormattedAxisData(
      testRecord,
      testScalarChannels,
      'timestamp'
    );
    expect(result).toEqual(unixTimestamp);
  });

  it('formats shot number correctly', () => {
    let result = getFormattedAxisData(
      testRecord,
      testScalarChannels,
      'activeExperiment'
    );
    expect(result).toEqual(testRecord.metadata.shotnum);

    testRecord.metadata.shotnum = undefined;
    result = getFormattedAxisData(testRecord, testScalarChannels, 'shotnum');
    expect(result).toEqual(NaN);
  });

  it('formats activeArea correctly', () => {
    const result = getFormattedAxisData(
      testRecord,
      testScalarChannels,
      'activeArea'
    );
    expect(result).toEqual(parseInt(testRecord.metadata.activeArea));
  });

  it('formats activeExperiment correctly', () => {
    testRecord.metadata.activeExperiment = '4';
    let result = getFormattedAxisData(
      testRecord,
      testScalarChannels,
      'activeExperiment'
    );
    expect(result).toEqual(parseInt(testRecord.metadata.activeExperiment));

    testRecord.metadata.activeExperiment = undefined;
    result = getFormattedAxisData(
      testRecord,
      testScalarChannels,
      'activeExperiment'
    );
    expect(result).toEqual(NaN);
  });

  it('formats channel data correctly', () => {
    let result = getFormattedAxisData(testRecord, testScalarChannels, 'test_3');
    expect(result).toEqual(
      (testRecord.channels['test_3'] as ScalarChannel).data
    );

    (testRecord.channels['test_3'] as ScalarChannel).data = '1';
    result = getFormattedAxisData(testRecord, testScalarChannels, 'test_3');
    expect(result).toEqual(1);

    result = getFormattedAxisData(
      testRecord,
      testScalarChannels,
      'invalid_channel'
    );
    expect(result).toEqual(NaN);
  });
});

describe('ConnectedPlot component', () => {
  let props: ConnectedPlotProps;

  const createView = () => {
    return renderComponentWithProviders(<ConnectedPlot {...props} />);
  };

  it('constructs data to insert into a Plot component', () => {
    props = {
      records: testRecords,
      channels: testScalarChannels,
      XAxis: 'timestamp',
      YAxis: 'shotnum',
      title: 'test title',
      type: 'scatter',
      XAxisSettings: {
        scale: 'time',
      },
      YAxesSettings: {
        scale: 'linear',
      },
    };

    const view = createView();
    expect(view.asFragment()).toMatchSnapshot();
  });

  it('inserts null points if an x or y value does not exist', () => {
    const amendedTestRecords = testRecords;
    amendedTestRecords[1].metadata.activeExperiment = undefined;
    props = {
      records: amendedTestRecords,
      channels: testScalarChannels,
      XAxis: 'activeArea',
      YAxis: 'activeExperiment',
      title: 'test title',
      type: 'scatter',
      XAxisSettings: {
        scale: 'linear',
      },
      YAxesSettings: {
        scale: 'log',
      },
    };

    const view = createView();
    expect(view.asFragment()).toMatchSnapshot();
  });

  it('constructs an empty dataset if we have no records available', () => {
    props = {
      records: [],
      channels: testScalarChannels,
      XAxis: 'timestamp',
      YAxis: 'shotnum',
      title: 'test title',
      type: 'line',
      XAxisSettings: {
        scale: 'time',
      },
      YAxesSettings: {
        scale: 'log',
      },
    };

    const view = createView();
    expect(view.asFragment()).toMatchSnapshot();
  });
});
