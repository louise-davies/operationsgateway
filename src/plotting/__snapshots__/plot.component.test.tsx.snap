// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plot component renders a canvas element with the correct attributes passed the correct props for a scatter plot 1`] = `
<DocumentFragment>
  <div
    style="flex: 1 0 0px; max-height: calc(100% - 38px); max-width: 100%;"
  >
    <canvas
      data-data="{\\"datasets\\":[{\\"label\\":\\"test_1\\",\\"data\\":[{\\"shotNum\\":0,\\"test_1\\":0}],\\"parsing\\":{\\"yAxisKey\\":\\"test_1\\",\\"xAxisKey\\":\\"test x-axis\\"},\\"borderColor\\":\\"#e31a1c\\",\\"backgroundColor\\":\\"#e31a1c\\"},{\\"label\\":\\"test_2\\",\\"data\\":[{\\"shotNum\\":1,\\"test_2\\":1}],\\"parsing\\":{\\"yAxisKey\\":\\"test_2\\",\\"xAxisKey\\":\\"test x-axis\\"},\\"borderColor\\":\\"#e31a1c\\",\\"backgroundColor\\":\\"#e31a1c\\"},{\\"label\\":\\"test_3\\",\\"data\\":[{\\"shotNum\\":2,\\"test_3\\":2}],\\"parsing\\":{\\"yAxisKey\\":\\"test_3\\",\\"xAxisKey\\":\\"test x-axis\\"},\\"borderColor\\":\\"#e31a1c\\",\\"backgroundColor\\":\\"#e31a1c\\"}]}"
      data-options="{\\"responsive\\":true,\\"maintainAspectRatio\\":false,\\"interaction\\":{\\"mode\\":\\"index\\",\\"intersect\\":false},\\"plugins\\":{\\"title\\":{\\"display\\":true,\\"text\\":\\"scatter plot\\"},\\"zoom\\":{\\"zoom\\":{\\"wheel\\":{\\"enabled\\":true},\\"pinch\\":{\\"enabled\\":true},\\"drag\\":{\\"enabled\\":false},\\"mode\\":\\"xy\\"},\\"pan\\":{\\"enabled\\":true,\\"mode\\":\\"xy\\"}}},\\"scales\\":{\\"x\\":{\\"type\\":\\"time\\",\\"time\\":{\\"displayFormats\\":{\\"millisecond\\":\\"HH:mm:ss:SSS\\",\\"second\\":\\"HH:mm:ss\\",\\"minute\\":\\"HH:mm\\",\\"hour\\":\\"d MMM ha\\",\\"day\\":\\"d MMM\\",\\"week\\":\\"d MMM\\",\\"month\\":\\"MMM yyyy\\",\\"quarter\\":\\"MMM yyyy\\",\\"year\\":\\"yyyy\\"},\\"tooltipFormat\\":\\"yyyy-MM-dd HH:mm:ss\\"}},\\"y\\":{\\"type\\":\\"linear\\",\\"display\\":true,\\"position\\":\\"left\\"},\\"y2\\":{\\"type\\":\\"linear\\",\\"display\\":false,\\"position\\":\\"right\\",\\"grid\\":{\\"drawOnChartArea\\":false}}}}"
      data-type="scatter"
      height="400"
      id="my-chart"
      width="400"
    />
  </div>
</DocumentFragment>
`;

exports[`Plot component updates data object correctly by settings opacity to 0 for lines that are hidden 1`] = `
<DocumentFragment>
  <div
    style="flex: 1 0 0px; max-height: calc(100% - 38px); max-width: 100%;"
  >
    <canvas
      data-data="{\\"datasets\\":[{\\"label\\":\\"test_1\\",\\"data\\":[{\\"shotNum\\":0,\\"test_1\\":0}],\\"parsing\\":{\\"yAxisKey\\":\\"test_1\\",\\"xAxisKey\\":\\"test x-axis\\"},\\"borderColor\\":\\"rgba(0,0,0,0)\\",\\"backgroundColor\\":\\"rgba(0,0,0,0)\\"},{\\"label\\":\\"test_2\\",\\"data\\":[{\\"shotNum\\":1,\\"test_2\\":1}],\\"parsing\\":{\\"yAxisKey\\":\\"test_2\\",\\"xAxisKey\\":\\"test x-axis\\"},\\"borderColor\\":\\"#e31a1c\\",\\"backgroundColor\\":\\"#e31a1c\\"},{\\"label\\":\\"test_3\\",\\"data\\":[{\\"shotNum\\":2,\\"test_3\\":2}],\\"parsing\\":{\\"yAxisKey\\":\\"test_3\\",\\"xAxisKey\\":\\"test x-axis\\"},\\"borderColor\\":\\"#e31a1c\\",\\"backgroundColor\\":\\"#e31a1c\\"}]}"
      data-options="{\\"responsive\\":true,\\"maintainAspectRatio\\":false,\\"interaction\\":{\\"mode\\":\\"index\\",\\"intersect\\":false},\\"plugins\\":{\\"title\\":{\\"display\\":true,\\"text\\":\\"scatter plot\\"},\\"zoom\\":{\\"zoom\\":{\\"wheel\\":{\\"enabled\\":true},\\"pinch\\":{\\"enabled\\":true},\\"drag\\":{\\"enabled\\":false},\\"mode\\":\\"xy\\"},\\"pan\\":{\\"enabled\\":true,\\"mode\\":\\"xy\\"}}},\\"scales\\":{\\"x\\":{\\"type\\":\\"time\\",\\"time\\":{\\"displayFormats\\":{\\"millisecond\\":\\"HH:mm:ss:SSS\\",\\"second\\":\\"HH:mm:ss\\",\\"minute\\":\\"HH:mm\\",\\"hour\\":\\"d MMM ha\\",\\"day\\":\\"d MMM\\",\\"week\\":\\"d MMM\\",\\"month\\":\\"MMM yyyy\\",\\"quarter\\":\\"MMM yyyy\\",\\"year\\":\\"yyyy\\"},\\"tooltipFormat\\":\\"yyyy-MM-dd HH:mm:ss\\"}},\\"y\\":{\\"type\\":\\"linear\\",\\"display\\":true,\\"position\\":\\"left\\"},\\"y2\\":{\\"type\\":\\"linear\\",\\"display\\":false,\\"position\\":\\"right\\",\\"grid\\":{\\"drawOnChartArea\\":false}}}}"
      data-type="scatter"
      height="400"
      id="my-chart"
      width="400"
    />
  </div>
</DocumentFragment>
`;

exports[`Plot component updates options object correctly 1`] = `
<DocumentFragment>
  <div
    style="flex: 1 0 0px; max-height: calc(100% - 38px); max-width: 100%;"
  >
    <canvas
      data-data="{\\"datasets\\":[{\\"label\\":\\"test_1\\",\\"data\\":[{\\"shotNum\\":0,\\"test_1\\":0}],\\"parsing\\":{\\"yAxisKey\\":\\"test_1\\",\\"xAxisKey\\":\\"test x-axis\\"},\\"borderColor\\":\\"#e31a1c\\",\\"backgroundColor\\":\\"#e31a1c\\"},{\\"label\\":\\"test_2\\",\\"data\\":[{\\"shotNum\\":1,\\"test_2\\":1}],\\"parsing\\":{\\"yAxisKey\\":\\"test_2\\",\\"xAxisKey\\":\\"test x-axis\\"},\\"borderColor\\":\\"#e31a1c\\",\\"backgroundColor\\":\\"#e31a1c\\"},{\\"label\\":\\"test_3\\",\\"data\\":[{\\"shotNum\\":2,\\"test_3\\":2}],\\"parsing\\":{\\"yAxisKey\\":\\"test_3\\",\\"xAxisKey\\":\\"test x-axis\\"},\\"borderColor\\":\\"#e31a1c\\",\\"backgroundColor\\":\\"#e31a1c\\"}]}"
      data-options="{\\"responsive\\":true,\\"maintainAspectRatio\\":false,\\"interaction\\":{\\"mode\\":\\"index\\",\\"intersect\\":false},\\"plugins\\":{\\"title\\":{\\"display\\":true,\\"text\\":\\"line plot\\"},\\"zoom\\":{\\"zoom\\":{\\"wheel\\":{\\"enabled\\":true},\\"pinch\\":{\\"enabled\\":true},\\"drag\\":{\\"enabled\\":false},\\"mode\\":\\"xy\\"},\\"pan\\":{\\"enabled\\":true,\\"mode\\":\\"xy\\"}}},\\"scales\\":{\\"x\\":{\\"type\\":\\"linear\\",\\"time\\":{\\"displayFormats\\":{\\"millisecond\\":\\"HH:mm:ss:SSS\\",\\"second\\":\\"HH:mm:ss\\",\\"minute\\":\\"HH:mm\\",\\"hour\\":\\"d MMM ha\\",\\"day\\":\\"d MMM\\",\\"week\\":\\"d MMM\\",\\"month\\":\\"MMM yyyy\\",\\"quarter\\":\\"MMM yyyy\\",\\"year\\":\\"yyyy\\"},\\"tooltipFormat\\":\\"yyyy-MM-dd HH:mm:ss\\"}},\\"y\\":{\\"type\\":\\"log\\",\\"display\\":true,\\"position\\":\\"left\\"},\\"y2\\":{\\"type\\":\\"linear\\",\\"display\\":false,\\"position\\":\\"right\\",\\"grid\\":{\\"drawOnChartArea\\":false}}}}"
      data-type="line"
      height="400"
      id="my-chart"
      width="400"
    />
  </div>
</DocumentFragment>
`;
